### PR TITLE
Urlhaus

### PR DESCRIPTION
**Pull request by Shakedp**
## Status
Ready

## Related Issues
https://github.com/demisto/etc/issues/16520

## Description
Integration with URLhaus

## Required version of Demisto
5.0.0

## Does it break backward compatibility?
   - No

## Must have
- [X] Tests
- [ ] Documentation (with link to it)
   pending approval: https://github.com/demisto/etc/issues/17817
- [x] Code Review

### CODE REVIEW

File: Integrations/URLHaus/URLHaus.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+
+def extract_zipped_buffer(buffer):
+    with io.BytesIO() as bio:
+        bio.write(buffer)
+        with zipfile.ZipFile(bio) as z:
+            return z.read(z.namelist()[0])
+
+
+def query_url_information(url):
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: Nice functions breakdown. Consider using a dictionary next time, easier to maintain in bigger integrations.

File: Integrations/URLHaus/URLHaus.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+    'Content-Type': 'application/x-www-form-urlencoded',
+    'Accept': 'application/json'
+}
+
+''' HELPER FUNCTIONS '''
+
+
+def http_request(method, command, data=None):
+    url = f'{API_URL}/{command}/'
+    demisto.info(f'{method} {url}')
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: We don't need this log here, maybe in debug and if so then add something more indicative

File: Integrations/URLHaus/URLHaus.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+        if blacklist == 'spamhaus_dbl':
+            if status.endswith('domain') or (status.startswith('abused') and compromised_is_malicious):
+                blacklist_appearances.append((blacklist, status))
+        elif status == 'listed':
+            blacklist_appearances.append((blacklist, None))
+
+    if len(blacklist_appearances) >= threshold:
+        description = ''
+        for appearance in blacklist_appearances:
+            if not appearance[1]:
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: What if `appearance[1]` is `0`,`''`? Might be safer to use `if appearance[1] is None:`

File: Integrations/URLHaus/URLHaus.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+                'Tags': url_information.get('tags', [])
+            }
+
+            payloads = []
+            for payload in url_information.get('payloads', []):
+                vt_data = payload.get('virustotal', None)
+                vt_information = None
+                if vt_data:
+                    vt_information = {
+                        'Result': float(vt_data['percent']),
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: `.get(...)` is better here and below

File: Integrations/URLHaus/URLHaus.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+                                                           COMPROMISED_IS_MALICIOUS)
+
+            ec['DBotScore']['Score'] = dbot_score
+            if dbot_score == 3:
+                ec['URL']['Malicious'] = {
+                    'Vendor': 'URLhaus',
+                    'Description': description
+                }
+
+            ec['URLhaus.URL(val.ID && val.ID === obj.ID)'] = urlhaus_data
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: Please change to ec[outputPaths['url']] = urlhaus_data

File: Integrations/URLHaus/URLHaus.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+                                                           COMPROMISED_IS_MALICIOUS)
+
+            ec['DBotScore']['Score'] = dbot_score
+            if dbot_score == 3:
+                ec['domain']['Malicious'] = {
+                    'Vendor': 'URLhaus',
+                    'Description': description
+                }
+
+            ec['URLhaus.Domain(val.Name && val.Name === obj.Name)'] = urlhaus_data
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: outputPaths['domain']]

File: Integrations/URLHaus/URLHaus.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+            })
+
+    except Exception:
+        demisto.debug(traceback.format_exc())
+        return_error('Failed getting domain data, please verify the arguments and parameters')
+
+
+def file_command():
+    hash = demisto.args()['hash']
+    hash_type = demisto.args()['hash_type'].lower()
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: should be value list and then no need for the check below

File: Integrations/URLHaus/URLHaus.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+    if hash_type not in ['md5', 'sha256']:
+        return_error('Only accepting MD5 or SHA256 hash types')
+
+    try:
+        file_information = query_payload_information(hash_type, hash).json()
+
+        if file_information['query_status'] == 'ok' and file_information['md5_hash']:
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: Where do we handle the sha256?

File: Integrations/URLHaus/URLHaus.py [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+                'FirstSeen': first_seen,
+                'LastSeen': last_seen,
+                'DownloadLink': file_information.get('urlhaus_download', ''),
+                'URL': file_information.get('urls', [])
+            }
+
+            virus_total_data = file_information.get('virustotal')
+            if virus_total_data:
+                urlhaus_data['VirusTotal'] = {
+                    'Percent': float(file_information.get('virustotal', {'percent': 0})['percent']),
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: Nice

File: Integrations/URLHaus/URLHaus.py [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+                    'Link': file_information.get('virustotal', {'link': ''})['link']
+                }
+
+            ec = {
+                'File': {
+                    'Size': urlhaus_data['Size'],
+                    'MD5': urlhaus_data['MD5'],
+                    'SHA256': urlhaus_data['SHA256']
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: use get

File: Integrations/URLHaus/URLHaus.py [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+                },
+                'URLhaus.File(val.MD5 && val.MD5 === obj.MD5)': urlhaus_data
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: outputPaths['file']

File: Integrations/URLHaus/URLHaus.py [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+            }
+
+            human_readable = tableToMarkdown(f'URLhaus reputation for {hash_type.upper()} : {hash}',
+                                             {
+                                                 'URLhaus link': urlhaus_data['DownloadLink'],
+                                                 'Signature': urlhaus_data['Signature'],
+                                                 'MD5': urlhaus_data['MD5'],
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: get

File: Integrations/URLHaus/URLHaus.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+                'Type': entryTypes['note'],
+                'ContentsFormat': formats['json'],
+                'Contents': file_information,
+                'HumanReadable': human_readable,
+                'HumanReadableFormat': formats['markdown'],
+                'EntryContext': ec
+            })
+        elif (file_information['query_status'] == 'ok' and not file_information['md5_hash']) or \
+                file_information['query_status'] == 'no_results':
+            human_readable = f'## URLhaus reputation for {hash_type.upper()} : {hash}\n' \
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: what happens here with sha256? can we have it?

File: Integrations/URLHaus/URLHaus.py [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+                'HumanReadable': f'No results for SHA256: {file_sha256}',
+                'HumanReadableFormat': formats['markdown']
+            })
+        else:
+            demisto.results({
+                'Type': entryTypes['error'],
+                'ContentsFormat': formats['text'],
+                'Contents': res.content
+            })
+    except json.JSONDecodeError:
```
Comment (hirshbergshachar) [inline Integrations/URLHaus/URLHaus.py]: This is how we save the sample? Why do we handle this flow with except?

File: Integrations/URLHaus/URLHaus.py [hunk 15]
```diff
@@ -0,0 +1,495 @@
...
+USE_SSL = not demisto.params().get('insecure', False)
+
+# Remove proxy if not set to true in params
+if not demisto.params().get('proxy'):
+    os.environ.pop('HTTP_PROXY', None)
+    os.environ.pop('HTTPS_PROXY', None)
+    os.environ.pop('http_proxy', None)
+    os.environ.pop('https_proxy', None)
+
+THRESHOLD = int(demisto.params().get('threshold', 1))
```
[the remark this answers is not in the document]
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: THRESHOLD indicates how many blacklists says it's malicious, I think that 1 is good for a default behaviour, don't you? It's out of three blacklists.

File: Integrations/URLHaus/URLHaus.py (hunk 1 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: Not sure what you mean..

File: Integrations/URLHaus/URLHaus.py (hunk 7 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: It is.

Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: You can use !file and insert other values

File: Integrations/URLHaus/URLHaus.py (hunk 8 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: We don't, it's just a distinction from cases where the status is 'ok' but there is no data;.

File: Integrations/URLHaus/URLHaus.py (hunk 7 above)
Comment (hirshbergshachar) [reply Integrations/URLHaus/URLHaus.py]: remove arg and check by length

File: TestPlaybooks/Test-URLhaus.yml [hunk 16]
```diff
@@ -0,0 +1,663 @@
+id: c48ce58e-79ec-460f-818e-8b41689453cd
+version: 17
+name: Test-URLhaus
+starttaskid: "0"
+tasks:
+  "0":
```
Comment (hirshbergshachar) [inline TestPlaybooks/Test-URLhaus.yml]: connect the download

File: Integrations/URLHaus/URLHaus.py (hunk 1 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: We decided to leave it for the next one.

File: Integrations/URLHaus/URLHaus.py (hunk 13 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: What do you mean?

File: Integrations/URLHaus/URLHaus.py [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+                'ContentsFormat': formats['json'],
+                'Contents': file_information,
+                'HumanReadable': human_readable,
+                'HumanReadableFormat': formats['markdown'],
+                'EntryContext': ec
+            })
+        elif (file_information['query_status'] == 'ok' and not file_information['md5_hash']) or \
+                file_information['query_status'] == 'no_results':
+            human_readable = f'## URLhaus reputation for {hash_type.upper()} : {hash}\n' \
+                f'No results!'
```
[the remark this answers is not in the document]
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: Added detailed description about blacklist appearance

File: Integrations/URLHaus/URLHaus.py (hunk 14 above)
Comment (Shakedp) [reply Integrations/URLHaus/URLHaus.py]: Changed

**Comment by meirwah** [comment]
@hirshbergshachar @Shakedp can we merge this?

**Comment by kirbles19** [comment]
@Shakedp @meirwah Done updating descriptions.

**Comment by meirwah** [comment]
10x @kirbles19 !!

**Comment by kirbles19** [comment]
Done editing the .yml and .md files.

**Pull request merged**